klippy/extras/probe_eddy_current.py

```diff
@@ -436,5 +446,252 @@ def start_probe_session(self, gcmd):
+            toolhead.dwell(.2)
+        toolhead.wait_moves()
+        # Wait for sample collection to finish
+        reactor = self.printer.get_reactor()
+        evttime = reactor.monotonic()
+        while move_times:
+            evttime = reactor.pause(evttime + .1)
+        sample_temp = sum(temps) / len(temps)
+        for i, data in enumerate(probe_samples):
+            freqs = [d[1] for d in data]
+            zvals = [d[2] for d in data]
+            avg_freq = sum(freqs) / len(freqs)
+            avg_z = sum(zvals) / len(zvals)
+            kin_z = i * .5 + .05 + kin_pos[2]
+            logging.info(
+                "Probe Values at Temp %.2fC, Z %.4fmm: Avg Freq = %.6f, "
+                "Avg Measured Z = %.6f"
+                % (sample_temp, kin_z, avg_freq, avg_z)
+            )
+            self.calibration_samples[i].append((sample_temp, avg_freq))
+        return sample_temp
+
+    def start_calibration(self):
+        self.enabled = False
+        self.calibration_samples = [[] for _ in range(DRIFT_SAMPLE_COUNT)]
+
+    def finish_calibration(self, success):
+        cal_samples = self.calibration_samples
+        self.calibration_samples = None
+        if not success:
+            return
+        gcode = self.printer.lookup_object("gcode")
+        if len(cal_samples) < 3:
+            raise gcode.error(
+                "calbration error, not enough samples"
+            )
+        min_temp, _ = cal_samples[0][0]
+        polynomials = []
+        for i, coords in enumerate(cal_samples):
+            height = .05 + i * .5
+            poly = Polynomial2d.fit(coords)
+            polynomials.append(poly)
+            logging.info("Polynomial at Z=%.2f: %s" % (height, repr(poly)))
+        self._check_calibration(polynomials, min_temp)
+        coef_cfg = "\n" + "\n".join([str(p) for p in polynomials])
+        configfile = self.printer.lookup_object('configfile')
+        configfile.set(self.name, "drift_calibration", coef_cfg)
+        configfile.set(self.name, "drift_calibration_min_temp", min_temp)
+        gcode.respond_info(
+            "%s: generated %d 2D polynomials\n"
+            "The SAVE_CONFIG command will update the printer config "
+            "file and restart the printer."
+            % (self.name, len(polynomials))
+        )
+        try:
+            # Dump collected data to temporary file
+            import json
+            ctime = int(self.printer.get_reactor().monotonic())
+            tmpfname = "/tmp/eddy-probe-drift-%d.json" % (ctime)
+            out = {
+                "polynomial_coefs": [c.get_coefs() for c in polynomials],
+                "legend": ["temperature", "frequency"],
+                "data": cal_samples,
+                "start_z": .05,
+                "sample_z_dist": .5
+            }
+            with open(tmpfname, "w") as f:
+                f.write(json.dumps(out))
+        except Exception:
+            logging.exception("Failed to write %s" % (tmpfname))
```

Comment:
I think we should avoid writing ad-hoc files from Klipper.  If this is just debug code, maybe we should remove it before merging in the main Klipper repo.  If it's generally useful information then I think we should look to export it via the Klipper API interface.

The ad-hoc files are hard for users to use, hard for us to document, and can become a burden to maintain as future devs wont necessarily know if they can change the contents.

Reply:
I'll remove it.  Its purpose was primarily for collecting data from other users testing the module.